packages/auth/src/providers/cognito/apis/resetPassword.ts

```
@@ -0,0 +1,55 @@
+import type { 
+	ForgotPasswordCommandOutput 
+} from '@aws-sdk/client-cognito-identity-provider';
+import { AuthError } from '../../../errors/AuthError';
+import { AuthValidationErrorCode } from '../../../errors/types/validation';
+import { assertServiceError } from '../../../errors/utils/assertServiceError';
+import { assertValidationError } from '../../../errors/utils/assertValidationError';
+import { 
+	AuthResetPasswordStep,
+	AuthStandardAttributeKey, 
+	DeliveryMedium, 
+	ResetPasswordRequest, 
+	ResetPasswordResult 
+} from '../../../types';
+import { CustomAttribute } from '../types/models/CustomAttribute';
+import { CognitoResetPasswordOptions } from '../types/options/CognitoResetPasswordOptions';
+import { resetPasswordClient } from '../utils/clients/ResetPasswordClient';
+
+export async function resetPassword(
+	resetPasswordRequest: ResetPasswordRequest<CognitoResetPasswordOptions>
+): Promise<ResetPasswordResult<AuthStandardAttributeKey | CustomAttribute>> {
+	const username: string = resetPasswordRequest.username;
+	assertValidationError(
+		!!username,
+		AuthValidationErrorCode.EmptyResetPasswordUsername
+	);
+	const config = Amplify.config;
+	try {
+		const res: ForgotPasswordCommandOutput = await resetPasswordClient({
+			Username: username,
+			ClientMetadata: 
+				resetPasswordRequest.options?.serviceOptions?.clientMetadata ?? 
+				config.clientMetadata
+		});
+		const codeDeliveryDetails = res.CodeDeliveryDetails;
+		return {
+			isPasswordReset: false,
+			nextStep: {
+				resetPasswordStep: AuthResetPasswordStep.CONFIRM_RESET_PASSWORD_WITH_CODE,
+				codeDeliveryDetails: {
+					deliveryMedium: codeDeliveryDetails?.DeliveryMedium as DeliveryMedium,
+					destination: codeDeliveryDetails?.Destination as string,
+					attributeName: codeDeliveryDetails?.AttributeName as AuthStandardAttributeKey,
+				},
+			},
+		};
+	} catch (error) {
+		assertServiceError(error);
+		throw new AuthError({ name: error.name, message: error.message });
+	}
```

I don't think we need to do it now, but looking and your & Israel's PRs it seems like this pattern would be repeated for all APIs. I wonder if it makes sense to push this down into the client itself (i.e. `UserPoolHttpClient`) and let the AuthError bubble up unless an API needed some unique error handling.

---

@jimblanc What you are saying is that instead of intercepting the missing username, we just let it go to the Cognito service and return the Cognito erorr?

I tend to agree with that. It seems nice to prevent a round trip, but trying to do 1-off validation handling like this could lead to inconsistencies.

---

My suggestion was more about just moving this "common" error handling down a level or two, since this try-catch is essentially just checking for unknown service errors + wrapping any errors in an AuthError. The same logic exists in [signUp](https://github.com/aws-amplify/amplify-js/pull/11153/files#diff-8b8e0c0eaa2c2b2add12d12aae99d2583f07e8a42f9d585bd3b80aad86c0aad8R105) and I imagine it would exist in the other APIs as well by default unless we needed to do some special error handling for particular APIs.

Regarding the pre-request validation, I think I'm ambivalent. I like preventing the unneeded round trip (which might have benefits as far as avoiding cognito throttling, extra network calls, etc.), but keeping client-side validation in sync with the service is always tricky. Now that I think about it though, I wonder if we can just lean on TypeScript to enforce this type of basic validation?

---

Argh, sorry @jimblanc I think I misread your comment (or rather applied your comment to the wrong block of code). Moving the wrapping of the service error into an AuthError seems like something that could definitely go into the client abstraction layer.

@helgabalashova Could you make that change for both resetPassword and signUp, if you agree with this?